tests/fixtures.hh

```diff
@@ -0,0 +1,259 @@
+    int new_rule_2;
+    std::vector<unsigned char> correct_indices_2;
+
+    double equivalent_minority;
+};
+
+
+class CapturedMapFixture {
+
+public:
+    CapturedMapFixture() : pmap(NULL), tree(NULL), root(NULL), num_not_captured(T_NNC),
+                         c(T_C), prediction(T_PRED), default_prediction(T_DEF_PRED),
+                         lower_bound(T_LOWER), objective(T_OBJ), len_prefix(4), equivalent_minority(T_MINOR) {
+        pmap = new CapturedPermutationMap();
+
+        tree = new CacheTree(T_NSAMPLES, T_NRULES, T_C, rules, labels, NULL, T_ABLATION, false, T_TYPE);
+
+        if(tree) {
+            tree->insert_root();
+            root = tree->root();
+        }
+
+        // First insertion
+        parent_prefix = {4, 2, 1};
+        new_rule = 5;
+
+        // Second insertion, for second and third tests
+        parent_prefix_2 = {1, 4, 5};
+        new_rule_2 = 2;
+
+        // Canonical prefix
+        std::vector<unsigned short> ordered_prefix = {1, 2, 4, 5};
+
+        rule_vinit(nsamples, &not_captured);
+
+        // Generate the not_captured vector, for use as the key in the captured map
+#ifdef GMP
+        // First find the captured vector, then take its complement
+        for(size_t i = 0; i < ordered_prefix.size(); i++) {
+            mpz_ior(not_captured, not_captured, rules[ordered_prefix.at(i)].truthtable);
+        }
+
+        mpz_com(not_captured, not_captured);
+#else
+        // Set all bits to 1 (since we won't be taking the complement as with GMP, initially
+        // having all bits as one means it initializes the vector to all being not_captured)
+        for(int i = 0; i < NENTRIES; i++) {
+            not_captured[i] = ~(not_captured[i] & 0);
+        }
+
+        for(size_t i = 0; i < ordered_prefix.size(); i++) {
+            for(int j = 0; j < NENTRIES; j++) {
+                not_captured[j] = not_captured[j] & ~(rules[ordered_prefix.at(i)].truthtable[j]);
+            }
+        }
+#endif
+    }
+
+    virtual ~CapturedMapFixture() {
+        rule_vfree(&not_captured);
+
+        if(pmap)
+            delete pmap;
+
+        if(tree)
+            delete tree;
+    }
+
+protected:
+
+    CapturedPermutationMap * pmap;
+    CacheTree * tree;
+    Node * root;
+
+    int num_not_captured;
+
+    double c;
+
+    bool prediction;
+    bool default_prediction;
+    double lower_bound;
+    double objective;
+    int len_prefix;
+
+    VECTOR not_captured;
+
+    tracking_vector<unsigned short, DataStruct::Tree> parent_prefix;
+    int new_rule;
+
+    tracking_vector<unsigned short, DataStruct::Tree> parent_prefix_2;
+    int new_rule_2;
+
+    double equivalent_minority;
+};
+
+
+class QueueFixture {
+
+public:
+    QueueFixture() : queue(NULL), tree(NULL), root(NULL) {
+        queue = new Queue(lb_cmp, "LOWER BOUND");
+
+        tree = new CacheTree(T_NSAMPLES, T_NRULES, T_C, rules, labels, NULL, T_ABLATION, false, T_TYPE);
```

Comment:
Why are you creating a tree in a queue fixture?

Reply:
Because in every test case where we're dealing with a queue, we also need to have nodes to push on that queue, and the tree provides us with an easy way of creating such nodes (instead of having to manually call the node constructor with a bunch of params, we can just do tree->insert_root).